qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/MainActivity.kt

```diff
@@ -62,74 +52,49 @@ class MainActivity : AppCompatActivity(), PermissionsListener {
         )
     }
 
-    override fun onExplanationNeeded(permissionsToExplain: List<String>) {
-        Toast
-            .makeText(
-                this,
-                "This app needs location and storage permissions" +
-                    "in order to show its functionality.",
-                Toast.LENGTH_LONG
-            ).show()
-    }
-
-    override fun onPermissionResult(granted: Boolean) {
-        if (granted) {
-            requestPermissionIfNotGranted(Manifest.permission.WRITE_EXTERNAL_STORAGE)
-        } else {
-            Toast.makeText(this, "You didn't grant location permissions.", Toast.LENGTH_LONG).show()
-        }
-    }
-
     override fun onRequestPermissionsResult(
         requestCode: Int,
         permissions: Array<String>,
         grantResults: IntArray
     ) {
         super.onRequestPermissionsResult(requestCode, permissions, grantResults)
-        if (requestCode == LOCATION_PERMISSIONS_REQUEST_CODE) {
-            permissionsHelper.onRequestPermissionsResult(requestCode, permissions, grantResults)
-        } else {
-
-            when (
-                grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED
-            ) {
-                true -> {
-                    binding.activitiesList.isClickable = true
-                }
-                else -> {
-                    binding.activitiesList.isClickable = false
-                    Toast.makeText(
-                        this,
-                        "You didn't grant storage or location permissions.",
-                        Toast.LENGTH_LONG
-                    ).show()
-                }
-            }
-        }
+        permissionsHelper.onRequestPermissionsResult(requestCode, permissions, grantResults)
     }
 
-    private fun requestPermissionIfNotGranted(permission: String) {
-        val permissionsNeeded = ArrayList<String>()
-
-        if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED
-        ) {
-            permissionsNeeded.add(permission)
-            ActivityCompat.requestPermissions(this, permissionsNeeded.toTypedArray(), 10)
+    private fun launchAfterPermissionResult(element: TestActivityDescription) {
+        lifecycleScope.launch {
+            permissionsHelper.checkAndRequestPermissions(
+                this@MainActivity,
+                DEFAULT_PERMISSIONS
+            )
+            element.launchActivityFun(this@MainActivity)
         }
     }
 
     private val activitySelectedDelegate:
-        GenericListAdapterItemSelectedFun<TestActivityDescription> = { positionAndElement ->
-            positionAndElement.second.launchActivityFun(this)
+        GenericListAdapterItemSelectedFun<TestActivityDescription> = { _, element ->
+            if (element.handlesPermissions) {
+                element.launchActivityFun(this)
+            } else {
+                launchAfterPermissionResult(element)
+            }
```

Comment:
This block is essentially the goal of this pull request. For any example, would like to conditionally decide if it should have location permissions requested before launching.

The launchAfterPermissionResult needs to block in a way that is attached to the `lifecycleScope`.